tools/behavior_simulator/src/state.rs

```diff
@@ -0,0 +1,280 @@
+                    (sender != player_number).then_some(IncomingMessage::Spl(*message))
+                })
+                .collect();
+            robot.database.main_outputs.game_controller_state = Some(GameControllerState {
+                game_state: GameState::Playing,
+                game_phase: GamePhase::Normal,
+                kicking_team: Team::Uncertain,
+                last_game_state_change: now,
+                penalties: Default::default(),
+                remaining_amount_of_messages: 1200,
+                set_play: None,
+            });
+            let messages = incoming_messages.iter().collect();
+            let messages = BTreeMap::from_iter(once((now, messages)));
+
+            robot.database.main_outputs.cycle_time.start_time = now;
+
+            robot.database.main_outputs.ball_position =
+                self.ball.as_ref().map(|ball| BallPosition {
+                    position: robot_to_field.inverse() * ball.position,
+                    last_seen: now,
+                });
+
+            robot.database.main_outputs.primary_state =
+                match (robot.penalized, self.filtered_game_state) {
+                    (true, _) => PrimaryState::Penalized,
+                    (false, FilteredGameState::Initial) => PrimaryState::Initial,
+                    (false, FilteredGameState::Ready { .. }) => PrimaryState::Ready,
+                    (false, FilteredGameState::Set) => PrimaryState::Set,
+                    (false, FilteredGameState::Playing { .. }) => PrimaryState::Playing,
+                    (false, FilteredGameState::Finished) => PrimaryState::Finished,
+                };
+            robot.database.main_outputs.filtered_game_state = Some(self.filtered_game_state);
+            robot.database.main_outputs.game_controller_state = Some(self.game_controller_state);
+
+            robot.cycle(messages)?;
+
+            for message in robot.interface.take_outgoing_messages() {
+                if let OutgoingMessage::Spl(message) = message {
+                    self.messages.push((*player_number, message));
+                    self.game_controller_state.remaining_amount_of_messages -= 1
+                }
+            }
+        }
+
+        if let Some(ball) = self.ball.as_mut() {
+            ball.position += ball.velocity * time_step.as_secs_f32();
+            ball.velocity *= 0.98;
+
+            if ball.position.x.abs() > 4.5 && ball.position.y < 0.75 {
+                events.push(Event::Goal);
+            }
+        }
+
+        self.time_elapsed += time_step;
+        self.cycle_count += 1;
+
+        Ok(events)
+    }
+
+    pub fn get_lua_state(&self) -> LuaState {
+        LuaState {
+            time_elapsed: self.time_elapsed.as_secs_f32(),
+            cycle_count: self.cycle_count,
+            // TODO: Expose robot data to lua again
+            // robots: self.robots.iter().map(LuaRobot::new).collect(),
+            robots: Default::default(),
```

Comment:
I would propose to move this to the next iteration.
Without providing full read/write access to robot databases, we have already a superset of features compared to the old simulator and I'm not sure yet how this may be done efficiently.